### PR TITLE
graph/path: add option to retain all paths from single-source path finding routines

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

### CODE REVIEW

File: graph/path/bellman_ford_moore.go [hunk 1]
```diff
@@ -78,6 +78,75 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
 	return path, true
 }
 
+// BellmanFordAllFrom returns a shortest-path tree for shortest paths from u to all nodes in
+// the graph g, or false indicating that a negative cycle exists in the graph. If the graph
+// does not implement Weighted, UniformCost is used.
+//
+// The time complexity of BellmanFordAllFrom is O(|V|.|E|).
```
Comment (romanwerpachowski) [inline graph/path/bellman_ford_moore.go:85]: Nitpicking: why not "*"?

File: graph/path/bellman_ford_moore.go [hunk 2]
```diff
@@ -78,6 +78,75 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
...
+
+	path = newShortestAltsFrom(u, nodes)
+	path.dist[path.indexOf[u.ID()]] = 0
+
+	// Queue to keep track which nodes need to be relaxed.
+	// Only nodes whose vertex distance changed in the previous iterations need to be relaxed again.
+	queue := newBellmanFordQueue(path.indexOf)
+	queue.enqueue(u)
+
+	// The maximum of edges in a graph is |V| * (|V|-1) which is also the worst case complexity.
```
Comment (romanwerpachowski) [inline graph/path/bellman_ford_moore.go:107]: "maximum number"?

File: graph/path/dijkstra.go [hunk 3]
```diff
@@ -84,6 +84,79 @@ func DijkstraFrom(u graph.Node, g traverse.Graph) Shortest {
...
 }
 
+// DijkstraAllFrom returns a shortest-path tree for shortest paths from u to all nodes in
+// the graph g. If the graph does not implement Weighted, UniformCost is used.
+// DijkstraAllFrom will panic if g has a u-reachable negative edge weight.
+//
+// If g is a graph.Graph, all nodes of the graph will be stored in the shortest-path
+// tree, otherwise only nodes reachable from u will be stored.
+//
+// The time complexity of DijkstrFrom is O(|E|.log|V|).
```
Comment (romanwerpachowski) [inline graph/path/dijkstra.go:94]: DijkstraAllFrom

File: graph/path/bellman_ford_moore.go (hunk 1 above)
Comment (kortschak) [reply graph/path/bellman_ford_moore.go:85]: Domain convention.

File: graph/path/dijkstra.go (hunk 3 above)
Comment (kortschak) [reply graph/path/dijkstra.go:94]: Fixed

File: graph/path/bellman_ford_moore.go (hunk 2 above)
Comment (kortschak) [reply graph/path/bellman_ford_moore.go]: Fixed

File: graph/path/bellman_ford_moore.go [hunk 4]
```diff
@@ -80,6 +80,77 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
...
+	// need to be relaxed again.
+	queue := newBellmanFordQueue(path.indexOf)
+	queue.enqueue(u)
+
+	// The maximum number of edges in a graph is |V| * (|V|-1) which is also
+	// the worst case complexity.
+	// If the queue-loop has more iterations than the amount of maximum edges
+	// it indicates that we have a negative cycle.
+	maxEdges := len(nodes) * (len(nodes) - 1)
+	var loops int
```
Comment (vladimir-ch) [inline graph/path/bellman_ford_moore.go:115]: nit: I would move this just before the for loop.

**Comment by kortschak** [comment]
This needs edits to bring it into line with the changes that were made in #1382.

File: graph/path/bellman_ford_moore.go (hunk 4 above)
Comment (kortschak) [reply graph/path/bellman_ford_moore.go:115]: OK. The rationale for it being where it is is to explain the reason for `loops`' existence. I'll send another PR after this one changing it here and in the original code after we have merged this.

**Comment by kortschak** [comment]
PTAL

File: graph/path/shortest.go [hunk 5]
```diff
@@ -174,6 +174,250 @@ func (p Shortest) To(vid int64) (path []graph.Node, weight float64) {
...
+	}
+
+	seen := make([]bool, len(p.nodes))
+	paths = p.allTo(from, to, seen, []graph.Node{p.nodes[to]}, nil)
+	weight = p.dist[to]
+
+	return paths, weight
+}
+
+func (p ShortestAlts) allTo(from, to int, seen []bool, path []graph.Node, paths [][]graph.Node) [][]graph.Node {
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:398]: This function is long enough to deserve a descriptive comment.

File: graph/path/shortest.go [hunk 6]
```diff
@@ -174,6 +174,250 @@ func (p Shortest) To(vid int64) (path []graph.Node, weight float64) {
...
+		c, ok := p.cycCosts[[2]int{mid, to}]
+		if !ok {
+			p.cycCosts[[2]int{mid, to}] = weight
+		} else if weight < c {
+			p.cycCosts[[2]int{mid, to}] = math.Inf(-1)
+		}
+	}
+}
+
+func (p ShortestAlts) addPath(to, mid int) {
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:280]: Can you add a comment saying what this function does?

File: graph/path/shortest.go [hunk 7]
```diff
@@ -174,6 +174,250 @@ func (p Shortest) To(vid int64) (path []graph.Node, weight float64) {
...
+	}
+	idx := len(p.nodes)
+	p.indexOf[uid] = idx
+	p.nodes = append(p.nodes, u)
+	p.dist = append(p.dist, math.Inf(1))
+	p.next = append(p.next, nil)
+	return idx
+}
+
+func (p ShortestAlts) set(to int, weight float64, mid int) {
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:267]: Can you add a comment saying what this function does?

File: graph/path/shortest.go [hunk 8]
```diff
@@ -174,6 +174,250 @@ func (p Shortest) To(vid int64) (path []graph.Node, weight float64) {
...
+	hasNegativeCycle bool
+
+	// cycCosts holds costs between
+	// pairs of nodes to report negative
+	// cycles. cycCosts is only used
+	// if it is non-nil.
+	cycCosts map[[2]int]float64
+}
+
+func newShortestAltsFrom(u graph.Node, nodes []graph.Node) ShortestAlts {
```
Comment (romanwerpachowski) [inline graph/path/shortest.go:223]: Can you add a comment saying what this function does?

File: graph/path/shortest.go (hunk 5 above)
Comment (kortschak) [reply graph/path/shortest.go:398]: I have made an issue to address these comments #1388; they exist on all three types, so I would prefer to have a single commit adding to all, distinct from this PR which is logic bearing.

Comment (kortschak) [reply graph/path/shortest.go:398]: #1389 is the draft, which I'll update for this PR's inclusions when it's merged.

**Comment by kortschak** [comment]
Yeah, I'll wait for @vladimir-ch since he reviewed (a long time ago) the original shortest path tree additions.

File: graph/path/bellman_ford_moore.go (hunk 4 above)
Comment (vladimir-ch) [reply graph/path/bellman_ford_moore.go:117]: If you made it on purpose, then it's fine to leave as is, of course.

**Pull request merged**